src/snapshot/mod.rs

```diff
@@ -0,0 +1,258 @@
+                            if others_total_vp != voting_power {
+                                acc.entry(last.0).or_default().push(KeyContribution {
+                                    reward_address,
+                                    value: voting_power - others_total_vp,
+                                });
+                            }
+                        }
+                    };
+                    acc
+                }),
+            stake_threshold,
+        }
+    }
+
+    pub fn stake_threshold(&self) -> Value {
+        self.stake_threshold
+    }
+
+    pub fn to_block0_initials(&self, discrimination: Discrimination) -> Initial {
+        Initial::Fund(
+            self.inner
+                .iter()
+                .map(|(vk, contribs)| {
+                    let value: Value = contribs.iter().map(|c| c.value).sum::<u64>().into();
+                    let address: Address =
+                        chain_addr::Address(discrimination, Kind::Account(vk.to_inner().into()))
+                            .into();
+                    InitialUTxO { address, value }
+                })
+                .collect::<Vec<_>>(),
+        )
+    }
+
+    pub fn voting_keys(&self) -> impl Iterator<Item = &Identifier> {
+        self.inner.keys()
+    }
+
+    pub fn contributions_for_voting_key<I: Into<Identifier>>(
+        &self,
+        voting_public_key: I,
+    ) -> Vec<KeyContribution> {
+        let voting_public_key: Identifier = voting_public_key.into();
+        self.inner
+            .get(&voting_public_key)
+            .cloned()
+            .unwrap_or_default()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use quickcheck::{Arbitrary, Gen};
+    use quickcheck_macros::*;
+    use std::num::NonZeroU64;
+
+    impl Arbitrary for RawSnapshot {
+        fn arbitrary<G: Gen>(g: &mut G) -> Self {
+            let n_registrations = usize::arbitrary(g);
+            Self(
+                (0..n_registrations)
+                    .map(|_| VotingRegistration::arbitrary(g))
+                    .collect::<Vec<_>>(),
+            )
+        }
+    }
+
+    #[quickcheck]
+    fn test_threshold(raw: RawSnapshot, stake_threshold: u64) {
+        let filtered_snapshot = raw
+            .0
+            .iter()
+            .filter(|reg| u64::from(reg.voting_power) >= stake_threshold)
+            .cloned()
+            .collect::<Vec<_>>();
+        assert_eq!(
+            Snapshot::from_raw_snapshot(raw.clone(), stake_threshold.into()),
+            Snapshot::from_raw_snapshot(
+                RawSnapshot(filtered_snapshot.clone()),
+                stake_threshold.into()
+            )
+        );
+        let mut snapshot = Snapshot::from_raw_snapshot(RawSnapshot(filtered_snapshot), 0.into());
+        snapshot.stake_threshold = stake_threshold.into();
+        assert_eq!(
+            Snapshot::from_raw_snapshot(raw, stake_threshold.into()),
+            snapshot
+        );
```

Comment:
I don't really see the point of this assertion. The `stake_threshold` in the argument could be 0 anyway, so in theory it should be covered too?

Reply:
correct, it was more of a double check that I did while writing this